### PR TITLE
[DOC] Enhancements for encoding.rdoc

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Adds sections:
- String Encoding
- Symbol and Regexp Encodings
- Filesystem Encoding
- Locale Encoding
- IO Encodings
  - External Encoding
  - Internal Encoding
- Script Encoding
- Transcoding
  - Transcoding a String

Subsection "Transcoding an IO Stream" is not ready.

### CODE REVIEW

**Comment by zverok** [comment]
@BurdetteLamar I have some rather prinipal problem with this approach to docs.

**Review by zverok — CHANGES_REQUESTED** [review]
@BurdetteLamar I have some rather principal problem with this approach to docs.
Currently, the preferred and most discoverable way of documenting things is this:
1. Considering Ruby's OO nature (and that almost any concept has corresponding class/module), most of the time, the "how the concept works" is documented in corresponding class
2. Documents in `doc/` are mostly covering very foundational-level concepts (syntax, variables etc), or some advanced concepts that are hard to attach to some module (security, refinements, extensions writing etc)

The problem is, documents in `doc/` are much less discoverable — at least in the current way of rendering of docs, be it https://docs.ruby-lang.org/en/master/ or https://ruby-doc.org/core-3.1.1/ — they don't have inherent structure and seem "auxiliary".

When somebody looks at "how encodings are handled in Ruby", they might look at strings or files, they might find a separate `Encoding` class in class list (or discover it themselves by experimenting in `irb` by just "`foo.encoding`, oh, interesting, what's this object?..")

Therefore, **at least until better documentation structure** (I am working on this currently), we should be very cautious about introducing "free-floating" files that just document some aspect—especially if the aspect is tightly coupled with existing classes.

For example, now we have documentation of encoding base concepts both in [Encoding](https://docs.ruby-lang.org/en/master/Encoding.html) class end `encoding.rdoc`, and if the user came to one of those from Google today, then to another one tomorrow, they would be confused (ugh, I remember I've seen the docs that were a bit different)—and it is very easy for docs to diverge in the future.

So, I have **two constructive proposals**:
1. Can we instead of doing what this PR does (and what #5572 did), improve class `Encoding` docs? And maybe check it is referred from where it is relevant (like IO and String). But we should make sure not to overexpand it: in base "what to do with encoding docs" links to corresponding methods is enough most of the time, there is no need to showcase _every_ concept in code, unless it is very foundational one
2. **This might've been feedback for #5572, but it was merged very soon. Can we please have a policy of docs merging that is somewhat less rushed? Especially when introducing new concepts, new files etc. Let's treat the docs the same as we treat code: if it is a simple fix, quick approve-and-merge is totally OK. But new file or completely rewritten docs for some core class wouldn't be hurt by some discussion and maybe two approvals?..** As far as I understand, the current policy is to merge everything that is grammatically and semantically correct, but I am very concerned with the general structure of understanding that is created by docs in total.

cc @jeremyevans

**Comment by jeremyevans** [comment]
In general, I'm fine with additional files in `doc`.  Such files should be linked to from the appropriate Ruby classes and methods.

Regarding encoding specifically, I think the separate file approach makes sense, because many Ruby programmers interact with encodings without explicitly using the `Encoding` class.  I think the `Encoding` class should reference the new `doc/encoding.rdoc` file, if it doesn't already.  `String#encode` would be another good place to link to the new file.

While I agree that not every concept needs an example, having examples is helpful.  There are a lot of programmers that learn best from examples.  One should also consider that many programmers may not be fluent in English, but are comfortable reading Ruby code.  The main cases where I discourage examples is for corner cases and other exceptional behavior (i.e. avoid showing examples of code that raises an exception), as well as multiple examples that show the same behavior (i.e. avoid redundancy).

I do think it makes sense to give additional time for review, especially now that we have more than one person doing reviews.  However, there needs to be a limit for how long one should be expected to wait. As a possible approach:

1. With 2 approvals, the PR can be merged right away
2. With 1 approval, please wait at least 2 days after submitting the PR before merging, to allow for additional review time.  You can have multiple PRs submitted simultaneously, so you don't need to wait for approval of one before working on the next.

**Comment by BurdetteLamar** [comment]
Second commit removed some stuff I thought might be wrong; third commit restored it.

**Comment by BurdetteLamar** [comment]
My aim in creating encoding.rdoc was to found a single document that makes sense of Ruby encoding, all in one place.  I'd planned on many links from elsewhere, which also would allow some material to be removed;  base concepts might be removed from Encoding, leaving mainily the methods documentation.  I didn't want to make this PR any larger.

Those links also make the document discoverable.  I'm thinking, for example, of the [dig document](https://docs.ruby-lang.org/en/master/doc/dig_methods_rdoc.html),
which links to its six relevant methods;  each method also links to the document.

**Comment by zverok** [comment]
@jeremyevans @BurdetteLamar Let me provide a bit more context of where I am coming from. (Probably PR comments is not the best place, but I'll make it all more visible in a few days/weeks, kinda distracted by war threat here in Ukraine)

I am [currently assigned the task](https://bugs.ruby-lang.org/issues/16512#change-96308) of providing better structured Ruby docs, with my [Ruby Reference](https://rubyreferences.github.io/rubyref/) used as a draft of the direction we want to move. What is currently rendered by RDoc by default (at docs.ruby-lang.org or alternative renderers) provides relatively handy reference for particular classes/methods, but the whole structure (especially in regards of contents of the `doc/` folder) is not really friendly, to say the least. 

**That's the reason** I am constantly thinking about the **big picture**, and how we can provide more consistent experience of "what's what" in Ruby, and "where's the answers".

One of my "guiding stars", so to say, is the **extreme object-oriented** nature of Ruby: the fact that most of the concepts are clearly mapped to particular classes. And I think we should keep it that way: most of "what we know about strings" is in `String` class docs, not in `docs/strings.rdoc`. It is clear and unanimous: if you google "ruby string" you'll be there, if you `Ctrl+F "string"` on a ruby docs index, you'll be there, if you `"foo".class` and look at its docs, you'll be there. The same is true for Array, Hash, numbers, etc., etc., etc. 

**I maintain the same is true for `Encoding`**: we have a separate named "encoding" concept, represented as a class. At every other points in docs it is easy to refer to there: You can just write at IO docs "...can be provided as an instance of Encoding or its name" and the rest will be done automatically by RDoc. You can google for "ruby encoding" and you should find single point of truth: `Encoding`. It is a feature, not a bug that everything encoding-related is in `Encoding`s docs. 

**We might** in the future separate docs in "brief reference style" vs "all the explanations style", but in the current and upcoming structure, **one point of truth** is important.

There are exceptions when something is **not** represented by named concept: `dig` is a good example (if we'll had module `Diggable`, the docs should've been there, but alas, we don't); some "deep dive" details like the recent command execution safety docs; etc. But **when we have a named concept**, we shouldn't duplicate it, ever.

@jeremyevans Additional answers:
> While I agree that not every concept needs an example, having examples is helpful. 

Yes, I didn't meant that we shouldn't provide examples at all. I just meant that "the more docs and examples, the better" is not an universal truth! We need to keep balance between "make it understandable" and "let quickly grasp a concept without reading 10 screens of detailed examples". To the "main" document (be it `doc/*.rdoc` or class-level docs) belong only the foundational examples to get the concept, detailed examples of each method's output are comfortable at the method-level docs.

> With 1 approval, please wait at least 2 days after submitting the PR before merging, to allow for additional review time.

I believe that "2 days" is critically small time span. I propose to treat docs as a code: we never merge Ruby feature in "no more than 2 days, if tests pass", right? Docs are important. Docs should be discussed, and might request change of structure, or different approach, or considering the whole picture. I have an utmost respect for the work done for increasing docs coverage,  but I have also a firm belief that "overall design" of the docs requires more collaboration.

**Comment by jeremyevans** [comment]
> I am [currently assigned the task](https://bugs.ruby-lang.org/issues/16512#change-96308) of providing better structured Ruby docs, with my [Ruby Reference](https://rubyreferences.github.io/rubyref/) used as a draft of the direction we want to move. What is currently rendered by RDoc by default (at docs.ruby-lang.org or alternative renderers) provides relatively handy reference for particular classes/methods, but the whole structure (especially in regards of contents of the `doc/` folder) is not really friendly, to say the least.

Could documentation organization be improved? Certainly.  However, I don't think what we have is bad, and it has been steadily improving.  We can always improve the organization further in the future.

> **I maintain the same is true for `Encoding`**: we have a separate named "encoding" concept, represented as a class. At every other points in docs it is easy to refer to there: You can just write at IO docs "...can be provided as an instance of Encoding or its name" and the rest will be done automatically by RDoc. You can google for "ruby encoding" and you should find single point of truth: `Encoding`. It is a feature, not a bug that everything encoding-related is in `Encoding`s docs.

When you deal with strings in Ruby, you are generally dealing directly with String instances.  When you are dealing with encodings in Ruby, you are often not dealing directly with Encoding instances.  This is an important distinction.

> > While I agree that not every concept needs an example, having examples is helpful.
> 
> Yes, I didn't meant that we shouldn't provide examples at all. I just meant that "the more docs and examples, the better" is not an universal truth! We need to keep balance between "make it understandable" and "let quickly grasp a concept without reading 10 screens of detailed examples". To the "main" document (be it `doc/*.rdoc` or class-level docs) belong only the foundational examples to get the concept, detailed examples of each method's output are comfortable at the method-level docs.

We should avoid documentation that is redundant.  That means avoiding both explanations that are redundant, as well as examples are redundant.  We should also avoid examples that solely offer proof of what the text says.  For example, the examples related to the default internal/external encodings are probably not necessary.  If we state `The default external encoding is UTF-8 for a text stream`, I don't think we benefit from an example showing that.  Examples should exist to help the reader understand a method/concept, not to prove that the text accompanying the example is correct.  So I agree with you that we could prune some of the examples in this PR without hurting the documentation.

> > With 1 approval, please wait at least 2 days after submitting the PR before merging, to allow for additional review time.
> 
> I believe that "2 days" is critically small time span. I propose to treat docs as a code: we never merge Ruby feature in "no more than 2 days, if tests pass", right? Docs are important. Docs should be discussed, and might request change of structure, or different approach, or considering the whole picture. I have an utmost respect for the work done for increasing docs coverage, but I have also a firm belief that "overall design" of the docs requires more collaboration.

I believe that approach to documentation is too slow.  In almost all cases, the new documentation being added is an improvement over what we have.  Is it perfect?  Of course not, nothing is.  However, as long as it is improvement on what we have, I don't think it makes sense taking weeks to wait for a second approval.  For well over a year, @BurdetteLamar has been merging after a single approval, so even waiting two days is a significant compromise.  You should understand that the more barriers there are to merging documentation PRs, the less documentation will be probably be written.

Note that if you request changes before merging, then @BurdetteLamar can work on the changes.  If you don't have time to review before merging, and the changes are merged without your review, you can always send your own pull requests to further improve the documentation.  Our focus should be on progressively improving the documentation, not trying to ensure that each documentation commit is perfect.

For example, if you want to later move the documentation from `doc/encoding.rdoc` to the `Encoding` class, we can discuss that.  As I mentioned, encoding in Ruby is often done without dealing directly with `Encoding` objects, so it's reasonable to have documentation on encoding in the `doc` folder.  However, as long as other places that deal with encoding can link to the documentation, it doesn't matter much where the documentation is located.  Please realize we can easily move documentation around later if we determine it will improve organization.

**Comment by duerst** [comment]
> > **I maintain the same is true for `Encoding`**: we have a separate named "encoding" concept, represented as a class. At every other points in docs it is easy to refer to there: You can just write at IO docs "...can be provided as an instance of Encoding or its name" and the rest will be done automatically by RDoc. You can google for "ruby encoding" and you should find single point of truth: `Encoding`. It is a feature, not a bug that everything encoding-related is in `Encoding`s docs.
> 
> When you deal with strings in Ruby, you are generally dealing directly with String instances. When you are dealing with encodings in Ruby, you are often not dealing directly with Encoding instances. This is an important distinction.

I fully agree with @jeremyevans here. For strings, String is the main thing. For encodings, Encoding isn't the main thing. It was introduced because occasionally, it comes in handy. But it shouldn't be at the center. People do a lot with encodings with Strings without ever even using an Encoding class.

> > I believe that "2 days" is critically small time span. I propose to treat docs as a code: we never merge Ruby feature in "no more than 2 days, if tests pass", right? Docs are important. Docs should be discussed, and might request change of structure, or different approach, or considering the whole picture. I have an utmost respect for the work done for increasing docs coverage, but I have also a firm belief that "overall design" of the docs requires more collaboration.
> 
> I believe that approach to documentation is too slow. In almost all cases, the new documentation being added is an improvement over what we have. Is it perfect? Of course not, nothing is. However, as long as it is improvement on what we have, I don't think it makes sense taking weeks to wait for a second approval. For well over a year, @BurdetteLamar has been merging after a single approval, so even waiting two days is a significant compromise. You should understand that the more barriers there are to merging documentation PRs, the less documentation will be probably be written.

I think in general, one approval should be fine. For bigger issues, it's fine to wait a bit longer, but that should be the exception, not the rule.

**Comment by BurdetteLamar** [comment]
So, where are we?

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:146]: This isn't correct.  If a string is given as an argument, then the returned string has the same encoding.  If you pass a literal string, you generally get UTF-8, because that is the default script encoding.

**Comment by zverok** [comment]
OK, so, first thing first.

@BurdetteLamar I think we can proceed with this file adding if both Jeremy and Martin are in favor of it. Please let me then review it in detail, I promise to do it in the next day or two, the comments would be cosmetic, OK?

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:192]: I don't think this is always true:

```ruby
:"\u1234".encoding
# => #<Encoding:UTF-8>
/\u1234/.encoding
# => #<Encoding:UTF-8>
```

I think the default encoding is US-ASCII if all characters are US-ASCII.  Otherwise, it depends on the script encoding:

```ruby
(eval "# encoding: iso-8859-1\n/\x87/").encoding
# => #<Encoding:ISO-8859-1>
(eval "# encoding: iso-8859-1\n:\200").encoding
# => #<Encoding:ISO-8859-1>
```

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:231]: Not sure if examples are helpful in this section.  Stating the defaults should be sufficient, the examples are much longer than the explanation and don't add a lot of value here.  We should try to avoid examples that attempt to prove what the text is stating is fact.

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:248]: This example is interesting, as it shows that IO.new doesn't inherit encoding. If you know how file descriptors work, this is not surprising, but it's probably worth mentioning the `IO.new` behavior separately.

**Comment by BurdetteLamar** [comment]
> OK, so, first thing first.
> 
> @BurdetteLamar I think we can proceed with this file adding if both Jeremy and Martin are in favor of it. Please let me then review it in detail, I promise to do it in the next day or two, the comments would be cosmetic, OK?

Ok, and thanks, Victor.  Take what time you need; I know there's a lot on your plate.

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:256]: In both this section and the next section, let's keep the File example, and remove the IO example.

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:282]: This example set is probably not needed.

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:297]: As above, let's keep the File example, and remove the IO example.

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:322]: s/encodingwith/encoding with/

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:346]: I don't think `b` transcodes.  `b` never changes the contents of the string, it is similar to `dup.force_encoding('BINARY')`.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks mostly good.  See inline comments.

**Comment by zverok** [comment]
Now, to more generic matters.

**(All of the below is not to offend Burdette Lamar or to say his work is not valuable. It is only tangential related to current PR, I am speaking of generical principles!)**

> @jeremyevans: Could documentation organization be improved? Certainly. However, I don't think what we have is bad, and it has been steadily improving. 

I do. (And, again, I am talking about the generic picture, not particular texts.)

If you'll look here: https://docs.ruby-lang.org/en/3.1/ from the point of view of the newcomer or, actually, any language user, and compare it to most of other popular language's `docs.lang-name.domain`, you'll see the difference: we have only two alphabetically-organized lists, of all `doc/*.rdoc` pages (even with folder structure like `doc/syntax/` lost), and of all classes of the language core AND the standard library. It is OK for going to a particular place you know, but absolutely not OK for any other use case (like get used to language or start studying it, or whatever). 

That's what I am concerned with, and that's where I am coming from with proposals of structuring stuff, of what should be related to what, etc.

> @jeremyevans: We can always improve the organization further in the future.

I don't think we are now in a state where we are missing a lot of docs, and they should be written & merged ASAP. We are in a state where particular aspects of documentation _could_ be improved here and there, but it is the overall structure that needs a critical amount of attention. 

> @jeremyevans: When you deal with strings in Ruby, you are generally dealing directly with String instances. When you are dealing with encodings in Ruby, you are often not dealing directly with Encoding instances. This is an important distinction. 

> @duerst: I fully agree with @jeremyevans here. For strings, String is the main thing. For encodings, Encoding isn't the main thing.

I see where you are coming from and agree to merge this PR, as said above. 

At the same time, I still believe that from the point of view of the documentation reader, a single point of truth is important (e.g. what you'll find in google by `Ruby Encoding`?..), and having the named class for the important concept can be seen as a possibility to create an entry point. `Encoding`-the-class is just easier to find, and easier to refer to from other docs (just literal `Encoding` would be highlighted by RDoc). Also, `Encoding` is what you'll be receiving when just experimenting in IRB in an attempt to find methods related; `Encoding` is where you'll refer when you need to know the encoding names supported.

I believe those might be seen as considerations overruling the fact that the user explicitly invokes the `Encoding` class less often than "does something about encodings". We'll get back to it at some point, I think. :wink: 

> @jeremyevans: I believe that approach to documentation is too slow. In almost all cases, the new documentation being added is an improvement over what we have. Is it perfect? Of course not, nothing is. However, as long as it is improvement on what we have, I don't think it makes sense taking weeks to wait for a second approval. For well over a year, @BurdetteLamar has been merging after a single approval, so even waiting two days is a significant compromise. You should understand that the more barriers there are to merging documentation PRs, the less documentation will be probably be written.

I am trying to establish the framework of discussion of the documentation as a consistent system, not just sum of paragraphs (with an implicit statement "the more, the better"). @BurdetteLamar has merge rights and I believe we work together quite fruitfully, so I am just asking that when we introduce _whole new files and sections_, we might discuss it a bit more, not some adamant rule that "at least N approvals, at least M days of discussions".

I am not for creating more entry barriers (I have my share of experience with maintaining OSS projects, and I think I have quite decent understanding of the balance between "not push away those who want to contribute" and "continue to maintain consistency").

To be honest, we don't have _any_ rush to merge a lot of docs at once:
1. The "deadline" for docs is as large as once-a-year (new version release), so we can comfortably have several PRs constructively discussed at once, nothing will be broken by this (obviously, if some newcomer wants to fix several typos, we can merge it at once, to encourage people to contribute!)
2. We (mostly) don't have a lot of totally missing docs that should be fixed ASAP (that was cared for through last years, by not a small part by people in this discussion)

The only thing I am trying to discuss is that we treat docs as not second-grade citizens, and be as vigilant and responsible as with code. With code, we might have some discussion of the PRs changing something significantly, right? And nobody will be unhappy about it, it is a normal process. And some PRs might be even rejected or rewritten completely in the process; when they don't fit in the complex system's consistency. 

>  Please realize we can easily move documentation around later if we determine it will improve organization.

In my experience, it is not as simple as that. 
The docs, especially for a large system, are a thing requiring constant maintenance. And the more docs we have, the more the maintenance cost. And the less attention we pay during _docs development_, the more the maintenance cost. For particular example in this PR, be some Encoding API changed in 3.2, somebody needs to remember to go through both `Encoding` class docs, and `encoding.rdoc`, and check examples and explanations validity, relevance, and fullness. I do as much as I can every December (and it takes several workdays), and still, a lot of things slip through.

That's another area I want to improve: to establish some documentation maintenance process. Some of it can be automated; some require constant human attention. The better and more consistent the structure of the docs, the more probability we'll keep them in the good shape.
And I am assured that "let's merge everything, we can always restructure it later" is not the most sustainable approach.

(Again, @BurdetteLamar, please don't take it all as a discussion of this particular PR or your work, I am just trying to establish some well-thought compromise)

**Comment by jeremyevans** [comment]
> If you'll look here: https://docs.ruby-lang.org/en/3.1/ from the point of view of the newcomer or, actually, any language user, and compare it to most of other popular language's `docs.lang-name.domain`, you'll see the difference: we have only two alphabetically-organized lists, of all `doc/*.rdoc` pages (even with folder structure like `doc/syntax/` lost), and of all classes of the language core AND the standard library. It is OK for going to a particular place you know, but absolutely not OK for any other use case (like get used to language or start studying it, or whatever).

It seems like the best way to fix that would be to add a documentation structure to the README, or at least a link in the README that takes you to a page with such a documentation structure. That documentation structure would link to related `doc` pages as needed (and potentially important classes), structured in a way that you think is best.  Is that something you would be interested in working on?

> That's what I am concerned with, and that's where I am coming from with proposals of structuring stuff, of what should be related to what, etc.
> 
> > @jeremyevans: We can always improve the organization further in the future.
> 
> I don't think we are now in a state where we are missing a lot of docs, and they should be written & merged ASAP. We are in a state where particular aspects of documentation _could_ be improved here and there, but it is the overall structure that needs a critical amount of attention.

I don't think anyone thinks we are in state where we are missing a lot of docs, so that sounds like a strawman to me.

Your comment about documentation being `written & merged ASAP` is not constructive.  @BurdetteLamar isn't rushing to "document all the things".  There is no rush, but nor should there be undue delay. 

These two documentation aspects are not mutually exclusive.  We _could_ benefit from more documentation structure, just as we _could_ benefit from improving the existing documentation.

> > @jeremyevans: I believe that approach to documentation is too slow. In almost all cases, the new documentation being added is an improvement over what we have. Is it perfect? Of course not, nothing is. However, as long as it is improvement on what we have, I don't think it makes sense taking weeks to wait for a second approval. For well over a year, @BurdetteLamar has been merging after a single approval, so even waiting two days is a significant compromise. You should understand that the more barriers there are to merging documentation PRs, the less documentation will be probably be written.
> 
> I am trying to establish the framework of discussion of the documentation as a consistent system, not just sum of paragraphs (with an implicit statement "the more, the better"). @BurdetteLamar has merge rights and I believe we work together quite fruitfully, so I am just asking that when we introduce _whole new files and sections_, we might discuss it a bit more, not some adamant rule that "at least N approvals, at least M days of discussions".

I encourage you to submit pull requests to implement your more consistent documentation system. 

Again, your comments `with an implicit statement "the more, the better"` are not constructive.  @BurdetteLamar is not writing documentation just to write documentation. 

You are welcome to submit a counter proposal for how review of new documentation files should be handled.  What does `a bit more` entail, in your opinion?

> 1. The "deadline" for docs is as large as once-a-year (new version release), so we can comfortably have several PRs constructively discussed at once, nothing will be broken by this (obviously, if some newcomer wants to fix several typos, we can merge it at once, to encourage people to contribute!)

My impression is that @BurdetteLamar prefers to have documentation reviewed one PR at a time.  As he is the one doing the majority of the work, I'm fine to tailor the review process to suit his workflow.

> The only thing I am trying to discuss is that we treat docs as not second-grade citizens, and be as vigilant and responsible as with code. With code, we might have some discussion of the PRs changing something significantly, right? And nobody will be unhappy about it, it is a normal process. And some PRs might be even rejected or rewritten completely in the process; when they don't fit in the complex system's consistency.

I agree with you about the importance of good documentation.  However, I don't think your approach of treating documentation PRs exactly the same as code PRs is a good idea.  Documentation is much easier to change/improve than code.  Once code is committed, people start relying on it, and you have to worry about backwards compatibility.  There is no need to worry about backwards compatibility for documentation, you can always improve it without worrying about whether your changes will break things.

There is a fundamental difference between documentation and code, and treating them the same is a bad idea.  That doesn't mean documentation is unimportant or a second-class citizen. It just means that the review process that works best for one does not necessarily work best for the other.

> > Please realize we can easily move documentation around later if we determine it will improve organization.
> 
> In my experience, it is not as simple as that. The docs, especially for a large system, are a thing requiring constant maintenance. And the more docs we have, the more the maintenance cost. And the less attention we pay during _docs development_, the more the maintenance cost. For particular example in this PR, be some Encoding API changed in 3.2, somebody needs to remember to go through both `Encoding` class docs, and `encoding.rdoc`, and check examples and explanations validity, relevance, and fullness. I do as much as I can every December (and it takes several workdays), and still, a lot of things slip through.

I agree that having to update documentation is a challenging issue that most large projects face.  I don't think our current approach involves paying less attention to the development of documentation.  Whether the documentation is located in the Encoding class or in a separate file that the encoding class links to doesn't really matter.  If you were reviewing the Encoding class documentation for changes, and come across a link to the separate encoding file, I would think you would review that file as well.  So the amount of work is roughly equal, in my opinion.

> That's another area I want to improve: to establish some documentation maintenance process. Some of it can be automated; some require constant human attention. The better and more consistent the structure of the docs, the more probability we'll keep them in the good shape. And I am assured that "let's merge everything, we can always restructure it later" is not the most sustainable approach.

If you have proposals to setup such a documentation maintenance process, please submit them (a feature request on Redmine would be best).

I don't believe this PR makes the current documentation structure worse or less consistent.  I think your description of our current approach as `let's merge everything, we can always restructure it later` is insulting, considering the amount of time that various reviewers (not just I) have spent reviewing documentation PRs and recommending improvements.

@zverok Please try to be constructive.  If you want more structured documentation, please submit a pull request for it and ask for review.  Repeatedly trashing our current documentation improvement process is not productive.

**Comment by zverok** [comment]
Sorry, I honestly didn't want to offend anybody and have highest respect to work done by people in this thread, abd other contributors.
I'll try to continue the constructive part of "structural sustainability" conversation in a separate tracker ticket after the war, probably.

**Comment by zverok** [comment]
@BurdetteLamar sorry for all the confusion and long thread.
Don't wait for my comments.

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: Done.

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: Done.

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: Done.

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: I have removed  the example, in keeping with changes elsewhere.  Doesn't this mention belong over at IO.new?

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: Done.

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: Removed.

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: Done.

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: Done.

File: doc/encoding.rdoc
Comment (BurdetteLamar) [reply doc/encoding.rdoc]: Done.

File: doc/encoding.rdoc
Comment (jeremyevans) [reply doc/encoding.rdoc]: I think it makes sense to mention this in the IO.new documentation.

File: doc/encoding.rdoc
Comment (jeremyevans) [inline doc/encoding.rdoc:233]: Now that we are only showing examples with `File`, I think these `# File` lines can be removed.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks good. See one inline comment.  After fixing that, OK to commit without further review.

**Comment by duerst** [comment]
I found the following; I'm not sure it's still relevant; if not, just ignore:
+_Transcoding_ is the process of revising the content of a string or stream
+by changing its encoding.

I would change that to read "Transcoding is the process of converting a string (i.e. a sequence of characters) from one encoding to another."
"revising the content" isn't clear at all.
You may add something like "As far as possible, the characters of the string stay the same, but the bytes that represent them may change."
In addition, it may be good to say something like "For characters that cannot be represented in the target encoding, several actions are possible, see options."

**Pull request merged**
